File: KengenWaterApp/app/src/main/java/com/water/alkaline/kengen/ui/activity/PreviewActivity.java
package com.water.alkaline.kengen.ui.activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.FileProvider;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.text.Html;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.water.alkaline.kengen.MyApplication;
import com.water.alkaline.kengen.R;
import com.water.alkaline.kengen.databinding.ActivityPreviewBinding;
import com.water.alkaline.kengen.databinding.DialogLoadingBinding;
import com.water.alkaline.kengen.library.ActionListeners;
import com.water.alkaline.kengen.library.ViewToImage;
import com.water.alkaline.kengen.model.SaveEntity;
import com.water.alkaline.kengen.placements.NativeAds;
import com.water.alkaline.kengen.utils.Constant;
import com.preference.PowerPreference;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class PreviewActivity extends AppCompatActivity {

    ActivityPreviewBinding binding;
    ArrayList<SaveEntity> mList = new ArrayList<>();
    int pos = 0;

    Dialog loaderDialog;

    public void dismiss_loader_dialog() {
        if (loaderDialog != null && loaderDialog.isShowing())
            loaderDialog.dismiss();
    }

    public void loader_dialog() {
        loaderDialog = new Dialog(this, R.style.NormalDialog);
        DialogLoadingBinding loadingBinding = DialogLoadingBinding.inflate(getLayoutInflater());
        loaderDialog.setContentView(loadingBinding.getRoot());
        Objects.requireNonNull(loaderDialog.getWindow()).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        loaderDialog.setCancelable(false);
        loaderDialog.setCanceledOnTouchOutside(false);
        loaderDialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        loaderDialog.show();
    }

    @Override
    protected void onResume() {
        super.onResume();
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                new NativeAds().shownative(PreviewActivity.this, null);
            }
        },250);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityPreviewBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        if (getIntent() != null && getIntent().hasExtra(Constant.POSITION)) {
            pos = getIntent().getIntExtra(Constant.POSITION, 0);
        }

        Glide.with(this).load(R.drawable.bg_splash).diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(binding.ivBG);

        try {
            Type type = new TypeToken<List<SaveEntity>>() {
            }.getType();

            mList = new Gson().fromJson(PowerPreference.getDefaultFile().getString(Constant.mList, new Gson().toJson(new ArrayList<SaveEntity>())), type);
            if (mList.get(mList.size() - 1).videoId.equalsIgnoreCase("99999"))
                mList.remove(mList.size() - 1);

        } catch (Exception e) {
            mList = new ArrayList<>();
        }

        Glide.with(this)
                .load(mList.get(pos).imgUrl)
                .placeholder(MyApplication.getPlaceHolder())
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(binding.frameContainer);

        binding.txtTitle.setText(Html.fromHtml(mList.get(pos).title));

        binding.includedToolbar.ivBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });

        binding.includedNative.adSpaceNative.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Constant.gotoAds(PreviewActivity.this);
            }
        });
        binding.btnStart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(PreviewActivity.this, PlayerActivity.class).putExtra(Constant.POSITION, pos));
            }
        });

        binding.btnShare.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                shareVideo();
            }
        });

    }



    public void shareVideo() {
        loader_dialog();
        String path = "";

        path = mList.get(pos).imgUrl;

        Glide.with(this).asBitmap().load(path)
                .into(new CustomTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(@NonNull @NotNull Bitmap resource, @Nullable @org.jetbrains.annotations.Nullable Transition<? super Bitmap> transition) {
                        saveBitmap(resource);
                    }

                    @Override
                    public void onLoadCleared(@Nullable @org.jetbrains.annotations.Nullable Drawable placeholder) {
                        Toast.makeText(PreviewActivity.this, "Something went Wrong", Toast.LENGTH_SHORT).show();
                        dismiss_loader_dialog();
                    }
                });
    }

    public void saveBitmap(Bitmap bitmap) {
        new ViewToImage(PreviewActivity.this, bitmap, new ActionListeners() {
            @Override
            public void convertedWithSuccess(Bitmap var1, String var2) {
                shareIImage(var2);
            }

            @Override
            public void convertedWithError(String var1) {
                Toast.makeText(PreviewActivity.this, "Something went Wrong", Toast.LENGTH_SHORT).show();
                dismiss_loader_dialog();
            }
        });
    }

    public void shareIImage(String mPath) {
        try {
            Intent i = new Intent(Intent.ACTION_SEND);
            i.setType("image/*");
            i.putExtra(Intent.EXTRA_SUBJECT, getResources().getString(R.string.app_name));

            String title = mList.get(pos).title;
            String sAux = PowerPreference.getDefaultFile().getString(Constant.vidShareMsg, "");

            String sAux2 = "https://play.google.com/store/apps/details?id=" + getPackageName();
            sAux = title + "\n\n" + sAux + "\n\n" + sAux2;

            Uri fileUri = FileProvider.getUriForFile(getApplicationContext(),
                    getPackageName() + ".fileprovider", new File(mPath));
            i.putExtra(Intent.EXTRA_TEXT, sAux);
            i.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            i.putExtra(Intent.EXTRA_STREAM, fileUri);
            startActivity(Intent.createChooser(i, "Choose One"));
            dismiss_loader_dialog();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(PreviewActivity.this, "Something went Wrong", Toast.LENGTH_SHORT).show();
            dismiss_loader_dialog();
        }

    }
}

File: KengenWaterApp/KengenWaterApp/app/src/main/java/com/water/alkaline/kengen/ui/activity/PreviewActivity.kt
package com.water.alkaline.kengen.ui.activity

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.text.Html
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.enableEdgeToEdge
import androidx.core.content.FileProvider
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.google.gms.ads.AdLoader
import com.google.gms.ads.MyApp
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.preference.PowerPreference
import com.water.alkaline.kengen.MyApplication
import com.water.alkaline.kengen.R
import com.water.alkaline.kengen.databinding.ActivityPreviewBinding
import com.water.alkaline.kengen.library.ActionListeners
import com.water.alkaline.kengen.library.ViewToImage
import com.water.alkaline.kengen.model.SaveEntity
import com.water.alkaline.kengen.ui.base.BaseActivity
import com.water.alkaline.kengen.utils.Constant
import com.water.alkaline.kengen.utils.onSingleClick
import com.water.alkaline.kengen.utils.uiController
import dagger.hilt.android.AndroidEntryPoint
import java.io.File

@AndroidEntryPoint
class PreviewActivity : BaseActivity() {

    private val binding by lazy {
        ActivityPreviewBinding.inflate(layoutInflater)
    }
    private var mList: ArrayList<SaveEntity> = ArrayList()
    private var pos: Int = 0

    override fun onResume() {
        super.onResume()
        if (MyApp.getAdModel().adsOnOff.equals("Yes", ignoreCase = true)) {
            if (binding.includedAd.flAd.childCount <= 0) {
                AdLoader.getInstance().showNativeSmall(this, binding.includedAd)
            }
        } else {
            binding.includedAd.cvAdMain.visibility = View.GONE
            binding.includedAd.flAd.visibility = View.GONE
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        if (intent != null && intent.hasExtra(Constant.POSITION)) {
            pos = intent.getIntExtra(Constant.POSITION, 0)
        }

        try {
            val type = object : TypeToken<List<SaveEntity?>?>() {
            }.type

            mList = Gson().fromJson(
                PowerPreference.getDefaultFile().getString(
                    Constant.mList, Gson().toJson(
                        ArrayList<SaveEntity>()
                    )
                ), type
            )

            if (mList[mList.size - 1].videoId.equals("99999", ignoreCase = true)) mList.removeAt(
                mList.size - 1
            )
        } catch (e: Exception) {
            mList = ArrayList()
        }

        if (mList.isNotEmpty()) {
            Glide.with(this)
                .load(mList[pos].imgUrl)
                .placeholder(MyApplication.getPlaceHolder())
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(binding.frameContainer)
            binding.txtTitle.text = Html.fromHtml(mList.get(pos).title)
            onClick()
            binding.btnStart.setOnClickListener {
                uiController.gotoIntent(
                    this,
                    Intent(
                        this@PreviewActivity,
                        PlayerActivity::class.java
                    ).putExtra(Constant.POSITION, pos),
                    true,
                    false
                )
            }
            binding.btnShare.setOnClickListener {
                shareVideo()
            }
        } else {
            Constant.showToast(this, "Unknown Error Occurred")
            finish()
        }
    }

    private fun onClick() {
        binding.includedToolbar.ivBack.setOnClickListener { onBackPressedDispatcher.onBackPressed() }
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                uiController.onBackPressed(this@PreviewActivity)
            }
        })
    }


    private fun shareVideo() {
        showLoadingDialog()
        var path: String? = ""
        path = mList[pos].imgUrl
        Glide.with(this).asBitmap().load(path)
            .into(object : CustomTarget<Bitmap?>() {
                override fun onResourceReady(
                    resource: Bitmap,
                    transition: Transition<in Bitmap?>?
                ) {
                    saveBitmap(resource)
                }

                override fun onLoadCleared(placeholder: Drawable?) {
                    hideLoadingDialog()
                }
            })
    }

    fun saveBitmap(bitmap: Bitmap?) {
        ViewToImage(this@PreviewActivity, bitmap, object : ActionListeners {
            override fun convertedWithSuccess(var1: Bitmap, var2: String) {
                hideLoadingDialog()
                shareIImage(var2)
            }

            override fun convertedWithError(var1: String) {
                Toast.makeText(this@PreviewActivity, "Something went Wrong", Toast.LENGTH_SHORT)
                    .show()
                hideLoadingDialog()
            }
        })
    }

    fun shareIImage(mPath: String?) {
        try {
            val i = Intent(Intent.ACTION_SEND)
            i.setType("image/*")
            i.putExtra(Intent.EXTRA_SUBJECT, resources.getString(R.string.app_name))

            val title = mList[pos].title
            var sAux = PowerPreference.getDefaultFile().getString(Constant.vidShareMsg, "")

            val sAux2 = "https://play.google.com/store/apps/details?id=" + packageName
            sAux = title + "\n\n" + sAux + "\n\n" + sAux2

            val fileUri = FileProvider.getUriForFile(
                applicationContext,
                packageName + ".fileprovider", File(mPath)
            )
            i.putExtra(Intent.EXTRA_TEXT, sAux)
            i.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            i.putExtra(Intent.EXTRA_STREAM, fileUri)
            startActivity(Intent.createChooser(i, "Choose One"))
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this@PreviewActivity, "Something went Wrong", Toast.LENGTH_SHORT).show()
        }
    }
}